lib/plugins/aws/customResources/resources/s3/handler.js

```diff
@@ -0,0 +1,85 @@
+'use strict';
+
+const { addPermission, removePermission } = require('./lib/permissions');
+const { updateConfiguration, removeConfiguration } = require('./lib/bucket');
+const { response, getEnvironment, getLambdaArn } = require('../utils');
+
+function handler(event, context) {
+  const PhysicalResourceId = 'CustomResouceExistingS3';
+  event = Object.assign({}, event, { PhysicalResourceId });
+
+  if (event.RequestType === 'Create') {
+    return create(event, context);
+  } else if (event.RequestType === 'Update') {
+    return update(event, context);
+  } else if (event.RequestType === 'Delete') {
+    return remove(event, context);
+  }
+  const error = new Error(`Unhandled RequestType ${event.RequestType}`);
+  return response(event, context, 'FAILED', {}, error);
+}
+
+function create(event, context) {
+  const { Region, AccountId } = getEnvironment(context);
+  const { FunctionName, BucketName, BucketConfig } = event.ResourceProperties;
+
+  const lambdaArn = getLambdaArn(Region, AccountId, FunctionName);
+
+  return addPermission({
+    functionName: FunctionName,
+    bucketName: BucketName,
+    region: Region,
+  })
+    .then(() =>
+      updateConfiguration({
+        lambdaArn,
+        region: Region,
+        functionName: FunctionName,
+        bucketName: BucketName,
+        bucketConfig: BucketConfig,
+      })
+    )
+    .then(() => response(event, context, 'SUCCESS'))
+    .catch(error => response(event, context, 'FAILED', {}, error));
+}
+
+function update(event, context) {
+  const { Region, AccountId } = getEnvironment(context);
+  const { FunctionName, BucketName, BucketConfig } = event.ResourceProperties;
+
+  const lambdaArn = getLambdaArn(Region, AccountId, FunctionName);
+
+  return updateConfiguration({
+    lambdaArn,
+    region: Region,
+    functionName: FunctionName,
+    bucketName: BucketName,
+    bucketConfig: BucketConfig,
+  })
+    .then(() => response(event, context, 'SUCCESS'))
+    .catch(error => response(event, context, 'FAILED', {}, error));
```

Comment:
What about generalizing a response handling to util? I think it'll reduce a significant amount of noise from endpoints here.

Do you plan to cover it with when adding support  for existing Cognito user pools?

Reply:
I like the idea but I'm not sure if I fully understand how this handling util would look like.

Can you provide any pointers?

Reply:
in `utils.js` we can expose handler which will work as handler decorator, e.g.

```javascript
modul.exports.awsHandler = main => {
  return (event, context, callback) => {
    Promise.resolve(main()).then(
      result => response(event, context, 'SUCCESS', result),
      error => response(event, context, 'FAILED', {}, error))
    ).then(result => callback(null, result), callback);
  }
}
```

And then in `handler.js` we can do:

```javascript
const { awsHandler } = require('./utils');
...
module.exports: = {
  handler: awsHandler(handler)
}
```

Reply:
Ah, that makes sense :+1:

Just implemented this and will ship it with the next PR update.